### PR TITLE
feat(loader): regex on_cmd expanded at generate time (closes #85)

### LINKED ISSUE #85: feat(loader): wildcard support in on_cmd (AOT-expanded at rvpm generate)
## Motivation

Users currently have to enumerate every user command a lazy plugin exposes in order to lazy-load it:

\`\`\`toml
[[plugins]]
url = \"xvzc/chezmoi.nvim\"
on_cmd = [\"ChezmoiEdit\", \"ChezmoiList\", \"ChezmoiEditWatch\", \"ChezmoiSourcePath\", ...]
\`\`\`

For plugins that expose a **prefix family** (\`Chezmoi*\`, \`GrugFar*\`, \`Rvpm*\`, \`Telescope*\`, \`Neotree*\`) this is:

- **Tedious** — users often miss secondary commands (\`ChezmoiSourcePath\`, \`GrugFarVisual\`) and realize only when the plugin fails to load.
- **Fragile** — new commands in the plugin's next release silently stop triggering the lazy load until the user re-reads the changelog.

Proposal: allow \`on_cmd = [\"Chezmoi*\"]\`, expanded at \`rvpm generate\` time by scanning the plugin's \`plugin/**/*.{lua,vim}\`.

### Precedent: lazy.nvim doesn't support this

Checked folke/lazy.nvim \`lua/lazy/core/handler/cmd.lua\` — it calls \`vim.api.nvim_create_user_command(cmd, …)\` once per string, no wildcard handling. Docs (\`lazy.folke.io/spec/lazy_loading\`) don't mention wildcards. No open issue requesting it.

Structurally lazy.nvim **can't** do AOT expansion — it is a pure runtime loader. A \`CmdUndefined\`-based runtime fallback is the only path there, and it **breaks \`:Prefix<Tab>\` completion** because \`CmdUndefined\` fires on invocation, not on completion. Probably why lazy.nvim skipped it.

rvpm has \`rvpm generate\` as an explicit AOT step (\`plugin/**/*.{vim,lua}\` already eagerly-globbed, phase markers baked, depends pre-pass done). Adding command scan there costs ~nothing at runtime and wins the completion UX.

### Completion behaviour

With AOT expansion each concrete command is registered as a stub via the existing \`nvim_create_user_command(…, { complete = fn })\` path. Neovim's built-in completion matches the stub list on \`:Chezmoi<Tab>\`, and the existing \`complete\` callback still loads the plugin on first Tab so the real plugin's completion takes over immediately. **Zero UX regression.**

### Config

\`\`\`toml
[[plugins]]
url = \"xvzc/chezmoi.nvim\"
on_cmd = [\"Chezmoi*\"]            # NEW: trailing-* glob, expanded at generate
\`\`\`

Mixed with exact names works too: \`on_cmd = [\"Chezmoi*\", \"ExtraDynamicCmd\"]\`. An exact name without \`*\` stays as-is (no scanning needed).

### Generate-time expansion

1. \`build_plugin_scripts\` already walks \`plugin/**/*.{vim,lua}\`. Collect command names on the same walk:
   - Lua:  \`vim.api.nvim_create_user_command%\([\\\"']([A-Z][A-Za-z0-9_]*)[\\\"']\`
   - Vim:  \`^\s*command!?\s+(?:-\S+\s+)*([A-Z][A-Za-z0-9_]*)\`
2. Attach as \`PluginScripts.defined_commands: Vec<String>\`.
3. In \`generate_loader\`, before the \`on_cmd\` loop, run \`expand_cmd_patterns(plugin_on_cmd, plugin.defined_commands)\`:
   - Entries containing \`*\` → prefix match against \`defined_commands\`, replace.
   - Non-wildcard entries pass through unchanged.
4. Feed the expanded list to the existing exact-match path. **loader.lua emission is unchanged.**

### Runtime behaviour

Identical to today. Each expanded command is a \`nvim_create_user_command\` stub with the existing bang/range/count/complete callback.

### Fallback / diagnostics

- Wildcard with zero matches → \`eprintln!(\"⚠ on_cmd pattern '{pat}' matched nothing for {plugin}; scan missed dynamically-defined commands?\")\` and keep the non-matching entries.
- \`rvpm doctor\` grows a check: \"plugins whose on_cmd wildcard pattern expanded to nothing\".

## Known limitations (to document)

- Only catches **statically-defined** commands. A plugin that defines commands via \`local name = vim.fn.input(); vim.api.nvim_create_user_command(name, …)\` (rare) will need explicit names.
- Only trailing \`*\` supported initially. \`?\`, \`[…]\`, leading/middle \`*\` YAGNI. Revisit if issues land.

## Plan (TDD)

1. \`expand_cmd_patterns(pats: &[String], defined: &[String]) -> Vec<String>\` pure fn + tests (glob match, exact passthrough, zero-match warn, duplicate dedup).
2. \`scan_defined_commands(files: &[PathBuf]) -> Vec<String>\` regex scan + tests (Lua \`nvim_create_user_command\`, vim-script \`command! [-opts] Name\`, mixed files, no false positives from strings in comments).
3. Wire \`PluginScripts.defined_commands\`, plumb through \`generate_loader\`.
4. Doctor check.
5. Docs: \`CLAUDE.md\` + README advanced section.

## Non-goals

- Dynamic runtime expansion via \`CmdUndefined\`. We'd inherit the lazy.nvim-skipped completion breakage. If a plugin's commands are un-scannable, users write exact names — same as today.
- Regex or full glob grammar. Trailing \`*\` only. Keep the config simple.

### PR DESCRIPTION
**Pull request by yukimemi**
Lets users write `on_cmd = ["/^Chezmoi/"]` instead of enumerating `["ChezmoiEdit", "ChezmoiList", ...]` by hand. `/regex/` entries are expanded at `rvpm generate` time against commands rvpm statically scans from the plugin's `plugin/`, `ftplugin/`, and `after/plugin/` files — runtime cost matches exact-name listing, `:Prefix<Tab>` completion stays intact (every matching stub is registered upfront).

## What's in

- **`src/cmd_scan.rs`** — new reusable module.
  - `scan_source(&str)` — regex-scan Lua `vim.api.nvim_create_user_command("Foo", …)` + vim-script `command![-opts]* Foo …`, filtered to Vim's uppercase-first-char rule (E183).
  - `scan_files(&[Path])` — aggregate + silently skip unreadable files.
  - `scan_plugin_commands(&Path)` — walk a plugin root's `plugin/` + `ftplugin/` + `after/plugin/` and aggregate. Exposed `pub` so the upcoming `rvpm add` auto-on_cmd feature can reuse it.
- **`loader::expand_cmd_patterns`** — takes the user's raw `on_cmd` list + the plugin's `defined_commands`, expands `/regex/`, preserves exact names, dedups. Invalid regex / zero matches → keep as literal + `⚠` warn so user notices.
- **`PluginScripts.defined_commands`** — populated in `build_plugin_scripts` via `cmd_scan::scan_plugin_commands`.
- **`generate_loader`** runs `expand_cmd_patterns` once before the existing stub emit loop. Zero Lua-side changes.
- New `regex = "1"` dep. ~400 KB, ~10 s compile time.

## Why regex over wildcards

User call: regex handles OR / suffix / mid-match without a second syntax. Vim-native `/.../` delimiter (mirroring `:s/.../` / `:/pat/`) is unambiguous against bare exact names.

For what it's worth, lazy.nvim doesn't support cmd wildcards at all (`lua/lazy/core/handler/cmd.lua` creates `nvim_create_user_command` per literal string, no expansion). Structurally it can't — pure runtime loader. rvpm has AOT `rvpm generate`, so this lands cleanly.

## Test plan

- `cmd_scan::tests` × 7 — Lua / Vim parse, dedup, uppercase filter, file aggregation, silent skip on missing file, walk scope.
- `loader::tests::expand_cmd_patterns_*` × 7 — exact pass, regex match, mixed, dedup with overlap, zero-match literal, invalid regex literal, `/` edge cases (`"/"`, `"/foo"`, `"foo/"` stay literal).
- Totals: 549 + 14 = 563 passed. clippy clean.
- Smoke: `rvpm generate` against real config produces loader.lua with unchanged exact-name stubs (no regression).

## Out of scope

- `rvpm add` integration using `scan_plugin_commands` — separate follow-up issue; helper is already `pub`.
- Regex flags / delimiter escapes (`/foo\//i`) — YAGNI until someone asks.

Closes #85.

## Summary by CodeRabbit

* **New Features**
  * `on_cmd` triggers now support regex patterns (e.g., `"/^Prefix.*/"`). Patterns are expanded during generation by scanning plugin command definitions, enabling efficient lazy-loading of multiple related commands while preserving command completion behavior.

* **Documentation**
  * Updated documentation to describe regex pattern syntax and behavior for command-based lazy-loading.

### CODE REVIEW

File: src/cmd_scan.rs [hunk 1]
```diff
@@ -0,0 +1,279 @@
...
+            }
+            if let Some(name) = extract_ident(rest)
+                && seen.insert(name.clone())
+            {
+                out.push(name);
+            }
+        }
+    }
+    out
+}
```
[the remark this answers is not in the document]
Comment (yukimemi) [reply src/cmd_scan.rs:77]: Thanks @gemini-code-assist — applied in 1204b77. `scan_source` now returns raw per-line hits in order; dedup lives in `scan_files` only. Updated the `scan_source_dedups_duplicates` test to `scan_source_preserves_duplicates` (asserts `["Foo", "Foo", "Foo"]`) and added `scan_files_dedups_across_sources` to pin the new contract.

File: src/loader.rs [hunk 2]
```diff
@@ -106,6 +112,70 @@ pub(crate) fn lua_quote(s: &str) -> String {
...
+                        if seen.insert(pat.clone()) {
+                            out.push(pat.clone());
+                        }
+                    } else {
+                        for m in matches {
+                            if seen.insert(m.clone()) {
+                                out.push(m.clone());
+                            }
+                        }
+                    }
```
[the remark this answers is not in the document]
Comment (yukimemi) [reply src/loader.rs]: Thanks @gemini-code-assist — applied in 1204b77 with `peekable()` + `peek()`, no `.collect()` anymore.

File: src/cmd_scan.rs [hunk 3]
```diff
@@ -0,0 +1,279 @@
...
+            let rest = &line[off..];
+            if let Some(open) = rest.find('(') {
+                let after = rest[open + 1..].trim_start();
+                if let Some(name) = extract_quoted_ident(after)
+                    && seen.insert(name.clone())
+                {
+                    out.push(name);
+                }
+            }
+        }
```
[the remark this answers is not in the document]
Comment (yukimemi) [reply src/cmd_scan.rs:54]: Thanks @coderabbitai — applied in 1204b77. Added `let lua_code = line.find("--").map_or(line, |i| &line[..i]);` before the Lua scan, so commented-out `nvim_create_user_command(...)` lines no longer contribute false stubs. Regression test `scan_source_ignores_lua_comment_out_definitions` pins it. Vim-script side was already fine (`" comment` fails the strip_prefix checks naturally).

File: src/loader.rs [hunk 4]
```diff
@@ -106,6 +112,70 @@ pub(crate) fn lua_quote(s: &str) -> String {
...
+                }
+                Err(e) => {
+                    eprintln!(
+                        "\u{26a0} on_cmd regex {pat:?} for {scope} failed to compile: {e}. \
+                         Kept as literal user command name."
+                    );
+                    if seen.insert(pat.clone()) {
+                        out.push(pat.clone());
+                    }
+                }
```
[the remark this answers is not in the document]
Comment (yukimemi) [reply src/loader.rs]: Thanks @coderabbitai — this was the serious one. Fixed in 1204b77: both the zero-match and invalid-regex branches now warn and **drop the entry** instead of re-emitting the raw pattern as a literal. Updated warn messages explain why. The old `_keeps_literal` / `_falls_back_to_literal` tests became `_drops_entry` (asserting `out.is_empty()`), plus a new `_drop_does_not_affect_valid_entries` pinning that other patterns in the same list keep expanding normally.
